config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java

```diff
@@ -689,6 +747,98 @@ protected void configure(HttpSecurity http) throws Exception {
 		}
 	}
 
+	@EnableWebSecurity
+	static class CustomJwtDecoderOnDsl extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer()
+						.jwt()
+							.decoder(new CustomJwtDecoder());
+			// @formatter:on
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderAsBean extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer();
+			// @formatter:on
+		}
+
+		@Bean
+		public JwtDecoder decoder() {
+			return new CustomJwtDecoder();
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderConflicting extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer()
+						.jwt()
+							.decoder(new CustomJwtDecoder());
+			// @formatter:on
+		}
+
+		@Bean
+		public JwtDecoder decoder() {
+			return new NimbusJwtDecoderJwkSupport("https://example.org");
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderAndJwkSetUri extends WebSecurityConfigurerAdapter {
+		@Value("${mock.jwk-set-uri:https://example.org}") String uri;
```

Comment:
I think we should be able to just hardcode this uri. This will make the test easier to follow.

Reply:
The trouble I ran into was that since it is using MockWebServer, then I don't know the uri at this point. However, I can see about instead auto-wiring MockWebServer into this class.

Reply:
Can we just fail to connect to the URL? I don't think we should be connecting to a server in tests.  Starting a server seems more like integration tests and should be done pretty sparingly rather than for edge cases.